### PR TITLE
Adding ensemble ungrib tasks

### PR DESCRIPTION
**Pull request by chunhuazhou**
## DESCRIPTION OF CHANGES: 
Adding changes for adding the ensemble forecast, starting from the ungrib tasks

### CODE REVIEW

File: workflow/exp/exp.ens_conus12km [hunk 1]
```diff
@@ -1,26 +1,66 @@
...
+
 export NET=rrfs                    # rrfs, rtma, 
 export MESH_NAME=conus12km         # conus12km, conus3km, atl3km, etc
-export VERSION=1.0.1
-export TAG=c1v01
-#
-export EXPDIR=${OPSROOT}/exp/${VERSION}/${EXP_NAME}
-export COMROOT=${OPSROOT}/${EXP_NAME}/com      # task input and output data as well as logs
-export DATAROOT=${OPSROOT}/${EXP_NAME}/stmp    # task workdirs ($DATA) which to be removed immediately upon task completion unless KEEPDATA=yes
+export WGF="ensf"   # working group function = "det", "enkf", "ensf", "firewx"
```
Comment (guoqing-noaa) [inline workflow/exp/exp.ens_conus12km:6]: @BenjaminBlake-NOAA @MatthewPyle-NOAA @hu5970 
I really think it is a good time to change `ensf` to `enfcst` as suggested by @BenjaminBlake-NOAA in https://github.com/NOAA-EMC/rrfs-workflow/pull/550#discussion_r1848239736.

rrfsv2 and rrfsv1 are different systems, we may not need to follow the misleading use of `ensf`

File: workflow/exp/exp.ens_conus12km [hunk 2]
```diff
@@ -1,26 +1,66 @@
...
+export LBC_LENGTH=3
+export LBC_INTERVAL=3
+export LBC_EXTRN_MDL_BASEDIR="${COMINgefs}/gefs.@Y@m@d/@H"
+export LBC_EXTRN_MDL_NAME_PATTERN="pgrb2ap5/gep#gmem#.t@Hz.pgrb2a.0p50.ffHHH"
+export LBC_EXTRN_MDL_NAME_PATTERN_b="pgrb2bp5/gep#gmem#.t@Hz.pgrb2b.0p50.ffHHH"
+#
+export RETRO_PERIOD='2024052623-2024052623'
+#
+export ACCOUNT="rtwrfruc"
+export ACCOUNT2="nrtrr"
```
Comment (guoqing-noaa) [inline workflow/exp/exp.ens_conus12km:54]: Do we use `ACCOUNT2` right now?

File: workflow/rocoto_funcs/ungrib_lbc.py [hunk 3]
```diff
@@ -8,16 +8,28 @@ def ungrib_lbc(xmlFile, expdir, do_ensemble=False):
   offset=int(os.getenv('LBC_OFFSET','6'))
   length=int(os.getenv('LBC_LENGTH','12'))
   interval=int(os.getenv('LBC_INTERVAL','3'))
+  fhr=os.getenv('FCST_LENGTH','6')
```
Comment (guoqing-noaa) [inline workflow/rocoto_funcs/ungrib_lbc.py:11]: The `fhr` variable is not used anywhere in this script.

File: workflow/exp/exp.ens_conus12km (hunk 2 above)
Comment (chunhuazhou) [reply workflow/exp/exp.ens_conus12km:54]: It is set in config_resources/config.jet and config_resources/config.hera but doesn't seem to be in the python scripts. I will remove it. Thanks @guoqing-noaa

Comment (chunhuazhou) [reply workflow/exp/exp.ens_conus12km:54]: Actually it is being used for clean task, as referenced in config_resources/config.base:
config_resources/config.base:export ACCOUNT_CLEAN=${ACCOUNT2:-$ACCOUNT}
Maybe we can just keep it?

Comment (guoqing-noaa) [reply workflow/exp/exp.ens_conus12km:54]: Thanks, @chunhuazhou 
We have this capability to use different ACCOUNT/QUEUE/etc for different tasks. But for now, we have NOT utilized that yet. We can add them later if there is a definite need to use different ACCOUNT/QUEUE etc.

Comment (guoqing-noaa) [reply workflow/exp/exp.ens_conus12km:54]: > Actually it is being used for clean task, as referenced in config_resources/config.base: config_resources/config.base:export ACCOUNT_CLEAN=${ACCOUNT2:-$ACCOUNT} Maybe we can just keep it?

Do we generate the clean task at this moment? If yes, we can keep it. Thanks!

Comment (chunhuazhou) [reply workflow/exp/exp.ens_conus12km]: Removed. We don't need to use it for now. Thanks @guoqing-noaa

File: workflow/rocoto_funcs/ungrib_lbc.py (hunk 3 above)
Comment (chunhuazhou) [reply workflow/rocoto_funcs/ungrib_lbc.py]: Removed. Thanks!

File: workflow/rocoto_funcs/ungrib_lbc.py [hunk 4]
```diff
@@ -8,16 +8,28 @@ def ungrib_lbc(xmlFile, expdir, do_ensemble=False):
   offset=int(os.getenv('LBC_OFFSET','6'))
   length=int(os.getenv('LBC_LENGTH','12'))
   interval=int(os.getenv('LBC_INTERVAL','3'))
+#  fhr=os.getenv('FCST_LENGTH','6')
```
Comment (guoqing-noaa) [inline workflow/rocoto_funcs/ungrib_lbc.py:11]: If we no longer use `fhr`, let's just remove line 11. We can always add it back if we have new needs.

File: workflow/rocoto_funcs/ungrib_lbc.py [hunk 5]
```diff
@@ -30,15 +42,15 @@ def ungrib_lbc(xmlFile, expdir, do_ensemble=False):
 '''
       meta_end=f'</metatask>\n'
       task_id=f'ungrib_lbc_g#group_index#'
+
+    if lbc_ungrib_group_total_num > 1:
+      dcTaskEnv['GROUP_INDEX'] = f'#group_index#'
+      dcTaskEnv['GROUP_TOTAL_NUM'] = f'{lbc_ungrib_group_total_num}'
+
```
Comment (hu5970) [inline workflow/rocoto_funcs/ungrib_lbc.py:49]: I just made a PR to remove this check.

File: jobs/JRRFS_UNGRIB [hunk 6]
```diff
@@ -27,26 +27,28 @@ export GROUP_TOTAL_NUM=${GROUP_TOTAL_NUM:-1}
 export pid=${pid:-$$}
 if [[ -z "${ENS_INDEX}" ]]; then
   export RUN="rrfs"
+  export SUBDIR=""
 else
   export RUN="ensrrfs"
+  export SUBDIR="/m${ENS_INDEX}"
 fi
```
Comment (hu5970) [inline jobs/JRRFS_UNGRIB:34]: We need to rethink how to make RUN and SUBDIR. Should add WGF as part of name, like

RUN="rrfs"${WGF}

**Comment by hu5970** [comment]
@chunhuazhou This PR needs to sync with PR 580 after it is merged. Thanks, Ming

File: workflow/exp/exp.ens_conus12km (hunk 1 above)
Comment (hu5970) [reply workflow/exp/exp.ens_conus12km:6]: We are now working on "enkf" even we will not add EnKF analysis step for now.

File: jobs/JRRFS_UNGRIB (hunk 6 above)
Comment (hu5970) [reply jobs/JRRFS_UNGRIB:34]: Also, should use WGF to decide how to set up SURDIR?

Comment (hu5970) [reply jobs/JRRFS_UNGRIB:34]: do we need SUBDIR?

File: workflow/exp/exp.ens_conus12km (hunk 1 above)
Comment (guoqing-noaa) [reply workflow/exp/exp.ens_conus12km:6]: But line 6 defines `ensf` instead of `enkf`. Should it be `enkf`?

File: jobs/JRRFS_UNGRIB (hunk 6 above)
Comment (chunhuazhou) [reply jobs/JRRFS_UNGRIB:34]: Yes, we do need it to have different directories for the members.

File: workflow/rocoto_funcs/ungrib_lbc.py (hunk 4 above)
Comment (chunhuazhou) [reply workflow/rocoto_funcs/ungrib_lbc.py]: Meant to remove it after testing but forgot to do so. Thanks @guoqing-noaa

File: workflow/exp/exp.ens_conus12km (hunk 1 above)
Comment (chunhuazhou) [reply workflow/exp/exp.ens_conus12km]: Changed it to enkf. Thanks!

File: workflow/exp/exp.ens_conus12km [hunk 7]
```diff
@@ -1,26 +1,64 @@
...
+export LBC_EXTRN_MDL_NAME_PATTERN_b="pgrb2bp5/gep#gmem#.t@Hz.pgrb2b.0p50.ffHHH"
+#
+export RETRO_PERIOD='2024052623-2024052623'
+#
+export ACCOUNT="rtwrfruc"
+export QUEUE="rth"
+export PARTITION=kjet
+export PARTITION2=kjet
+export RESERVATION="rrfsens"
+export RESERVATION2=""
```
[the remark this answers is not in the document]
Comment (chunhuazhou) [reply workflow/exp/exp.ens_conus12km]: Removed. Thanks @guoqing-noaa

File: workflow/exp/exp.ens_conus12km (hunk 1 above)
Comment (guoqing-noaa) [reply workflow/exp/exp.ens_conus12km]: Haha, this demonstrated one can easily get confused between `enkf` and `ensf`

**Review by hu5970 — APPROVED** [review]

File: workflow/exp/exp.ens_conus12km (hunk 1 above)
Comment (chunhuazhou) [reply workflow/exp/exp.ens_conus12km]: If we are not limited to 4 characters, ensda and enfcst (or ensfcst) makes more sense to me.

**Review by MatthewPyle-NOAA — APPROVED** [review]

**Pull request merged**